### ISSUE TITLE
paddle2.0-rc的LSTM网络有问题还是我写错了呢？

### ISSUE THREAD

**Issue by wkzqn** [PaddlePaddle/Paddle#28537]
![image](https://user-images.githubusercontent.com/44158381/98775404-0c626980-2428-11eb-834f-efb67d377158.png)

**Comment by wangxicoding** [comment]
我试了下你代码，cpu和gpu都可以正常运行，估计是环境的问题，能贴一下具体的core吗

**Comment by wkzqn** [comment]
稍等，我再检查下哈

**Comment by wkzqn** [comment]
import numpy as np
import paddle
import paddle.fluid as fluid
from paddle.fluid.dygraph import Linear, to_variable
from paddle.fluid.layers import concat
from paddle.nn import LSTM


class D(fluid.dygraph.Layer):
    def __init__(self):
        super(D, self).__init__()
        self.lstm = LSTM(100, 100)
        self.fc = Linear(input_dim=1 * 100, output_dim=1)

    def forward(self, x):
        _, (h, c) = self.lstm(x)
        x = h[-1, :, :]
        x = self.fc(x)
        return x

class G(fluid.dygraph.Layer):
    def __init__(self):
        super(G, self).__init__()
        self.lstm = LSTM(100, 100)

    def forward(self, z):
        _, (h, c) = self.lstm(z)
        x = h[-1, :, :]
        return x

use_gpu = True
place = fluid.CUDAPlace(0) if use_gpu else fluid.CPUPlace()
with fluid.dygraph.guard(place):
    g = G()
    d = D()
    d_opt = fluid.optimizer.AdamOptimizer(learning_rate=0.01,
                                        parameter_list=d.parameters())
    g_opt = fluid.optimizer.AdamOptimizer(learning_rate=0.01,
                                        parameter_list=g.parameters())
    x = paddle.randn((64, 10, 100))
    y = paddle.randn((64, 1, 100))
    x_real = concat(input=[x, y], axis=1)
    print(1)
    y_hat = g(x)
    print("y hat shape:", y_hat.shape)
    y_hat = y_hat.detach()
    y_hat = y_hat.reshape([y_hat.shape[0], 1, y_hat.shape[1]])
    print("y hat shape:", y_hat.shape)
    print(2)
    fake_loss = fluid.layers.mean(d(y_hat))
    real_loss = -1 * fluid.layers.mean(d(y))
    print(3)
    d_loss = fake_loss + real_loss
    print(4)
    d_loss.backward()
    print(5)
    d_opt.minimize(d_loss)
    print(6)
    d.clear_gradients()
    print(7)
    y_hat = g(x)
    y_hat = y_hat.reshape([y_hat.shape[0], 1, y_hat.shape[1]])
    g_loss = -1 * fluid.layers.mean(d(y_hat))
    g_loss.backward()
    g_opt.minimize(g_loss)
    g.clear_gradients()
    print(8)

**Comment by wkzqn** [comment]
`import numpy as np
import paddle
import paddle.fluid as fluid
from paddle.fluid.dygraph import Linear, to_variable
from paddle.fluid.layers import concat
from paddle.nn import LSTM


class D(fluid.dygraph.Layer):
    def __init__(self):
        super(D, self).__init__()
        self.lstm = LSTM(100, 100)
        self.fc = Linear(input_dim=1 * 100, output_dim=1)

    def forward(self, x):
        _, (h, c) = self.lstm(x)
        x = h[-1, :, :]
        x = self.fc(x)
        return x

class G(fluid.dygraph.Layer):
    def __init__(self):
        super(G, self).__init__()
        self.lstm = LSTM(100, 100)

    def forward(self, z):
        _, (h, c) = self.lstm(z)
        x = h[-1, :, :]
        return x

use_gpu = True
place = fluid.CUDAPlace(0) if use_gpu else fluid.CPUPlace()
with fluid.dygraph.guard(place):
    g = G()
    d = D()
    d_opt = fluid.optimizer.AdamOptimizer(learning_rate=0.01,
                                        parameter_list=d.parameters())
    g_opt = fluid.optimizer.AdamOptimizer(learning_rate=0.01,
                                        parameter_list=g.parameters())
    x = paddle.randn((64, 10, 100))
    y = paddle.randn((64, 1, 100))
    x_real = concat(input=[x, y], axis=1)
    print(1)
    y_hat = g(x)
    print("y hat shape:", y_hat.shape)
    y_hat = y_hat.detach()
    y_hat = y_hat.reshape([y_hat.shape[0], 1, y_hat.shape[1]])
    print("y hat shape:", y_hat.shape)
    print(2)
    fake_loss = fluid.layers.mean(d(y_hat))
    real_loss = -1 * fluid.layers.mean(d(y))
    print(3)
    d_loss = fake_loss + real_loss
    print(4)
    d_loss.backward()
    print(5)
    d_opt.minimize(d_loss)
    print(6)
    d.clear_gradients()
    print(7)
    y_hat = g(x)
    y_hat = y_hat.reshape([y_hat.shape[0], 1, y_hat.shape[1]])
    g_loss = -1 * fluid.layers.mean(d(y_hat))
    g_loss.backward()
    g_opt.minimize(g_loss)
    g.clear_gradients()
    print(8)`

**Comment by wkzqn** [comment]
```
import numpy as np
import paddle
import paddle.fluid as fluid
from paddle.fluid.dygraph import Linear, to_variable
from paddle.fluid.layers import concat
from paddle.nn import LSTM


class D(fluid.dygraph.Layer):
    def __init__(self):
        super(D, self).__init__()
        self.lstm = LSTM(100, 100)
        self.fc = Linear(input_dim=1 * 100, output_dim=1)

    def forward(self, x):
        _, (h, c) = self.lstm(x)
        x = h[-1, :, :]
        x = self.fc(x)
        return x

class G(fluid.dygraph.Layer):
    def __init__(self):
        super(G, self).__init__()
        self.lstm = LSTM(100, 100)

    def forward(self, z):
        _, (h, c) = self.lstm(z)
        x = h[-1, :, :]
        return x

use_gpu = True
place = fluid.CUDAPlace(0) if use_gpu else fluid.CPUPlace()
with fluid.dygraph.guard(place):
    g = G()
    d = D()
    d_opt = fluid.optimizer.AdamOptimizer(learning_rate=0.01,
                                        parameter_list=d.parameters())
    g_opt = fluid.optimizer.AdamOptimizer(learning_rate=0.01,
                                        parameter_list=g.parameters())
    x = paddle.randn((64, 10, 100))
    y = paddle.randn((64, 1, 100))
    x_real = concat(input=[x, y], axis=1)
    print(1)
    y_hat = g(x)
    print("y hat shape:", y_hat.shape)
    y_hat = y_hat.detach()
    y_hat = y_hat.reshape([y_hat.shape[0], 1, y_hat.shape[1]])
    print("y hat shape:", y_hat.shape)
    print(2)
    fake_loss = fluid.layers.mean(d(y_hat))
    real_loss = -1 * fluid.layers.mean(d(y))
    print(3)
    d_loss = fake_loss + real_loss
    print(4)
    d_loss.backward()
    print(5)
    d_opt.minimize(d_loss)
    print(6)
    d.clear_gradients()
    print(7)
    y_hat = g(x)
    y_hat = y_hat.reshape([y_hat.shape[0], 1, y_hat.shape[1]])
    g_loss = -1 * fluid.layers.mean(d(y_hat))
    g_loss.backward()
    g_opt.minimize(g_loss)
    g.clear_gradients()
    print(8)
```

**Comment by wkzqn** [comment]
@wangxicoding 你看下我这个代码呢？刚才那个代码我试了，确实gpu和cpu都可以运行，我这个代码只能在cpu上运行。我想做的是要实现一个gan网络，但是里面网络想要用lstm网络，但是现在我无法发现我哪里出了问题。

**Comment by wkzqn** [comment]
@wangxicoding 能再帮我看下吗？

**Comment by wkzqn** [comment]
第一次贴的代码有点问题，`with fluid.dygraph.guard(place)`这句代码写成了`with fluid.dygraph.guard(fluid.CPUPlace()):`，所以会一直在cpu上运行，我重新贴一下正确的代码，这次在gpu上运行则会报错了。：
```
import paddle
import paddle.fluid as fluid
from paddle.nn import LSTM

class G(fluid.dygraph.Layer):
    def __init__(self):
        super(G, self).__init__()
        self.lstm = LSTM(100, 100)

    def forward(self, z):
        _, (h, c) = self.lstm(z)
        x = h[-1, :, :]
        return x

use_gpu = True
place = fluid.CUDAPlace(0) if use_gpu else fluid.CPUPlace()
with fluid.dygraph.guard(place):
    g = G()
    g_opt = fluid.optimizer.AdamOptimizer(learning_rate=0.001, parameter_list=g.parameters())
    x = paddle.randn((64, 10, 100))
    print(1)
    g_loss = -1 * fluid.layers.mean(g(x))
    print(2)
    g_loss.backward()
    g_opt.minimize(g_loss)
    g.clear_gradients()
    print(3)
```
@wangxicoding

**Comment by wkzqn** [comment]
感觉确实有问题啊，这个最简单的demo都不行：
```
import paddle
import paddle.fluid as fluid
from paddle.nn import LSTM

use_gpu = True
place = fluid.CUDAPlace(0) if use_gpu else fluid.CPUPlace()
with fluid.dygraph.guard(place):
    rnn = paddle.nn.LSTM(100, 100)
    opt = fluid.optimizer.AdamOptimizer(learning_rate=0.001, parameter_list=rnn.parameters())
    x = paddle.randn((64, 10, 100))
    y, (h, c) = rnn(x)
    x = h[-1, :, :]  # x shape is: [64, 100]
    loss = -1 * fluid.layers.mean(x)
    loss.backward()
    opt.minimize(loss)
    rnn.clear_gradients()
    print("成功")
```
@wangxicoding

**Comment by wangxicoding** [comment]
``` python
import paddle
import paddle.fluid as fluid
from paddle.nn import LSTM

class G(fluid.dygraph.Layer):
    def __init__(self):
        super(G, self).__init__()
        self.lstm = LSTM(100, 100)

    def forward(self, z):
        _, (h, c) = self.lstm(z)
        x = h[-1, :, :]
        return x

use_gpu = True
place = fluid.CUDAPlace(0) if use_gpu else fluid.CPUPlace()
with fluid.dygraph.guard(place):
    g = G()
    g_opt = fluid.optimizer.Adam(learning_rate=0.001, parameter_list=g.parameters())
    x = paddle.randn((64, 10, 100))
    print(1)
    g_loss = -1 * fluid.layers.mean(g(x))
    print(g_loss)
    print(2)
    g_loss.backward()
    g_opt.minimize(g_loss)
    g.clear_gradients()
    print(3)
```
你前面的代码，可以跑呢。
![image](https://user-images.githubusercontent.com/10208305/98793721-3d03cc80-2443-11eb-8029-67eae9fe857d.png)

@wkzqn 把你错误log发一下~~

**Comment by wkzqn** [comment]
![image](https://user-images.githubusercontent.com/44158381/98801136-51989280-244c-11eb-8d16-95c52a801701.png)
我是在AI studio上跑的 @wangxicoding
